### ISSUE TITLE
📎 TS Parser Conformance

### ISSUE THREAD

**Issue by MichaReiser** [rome/tools#2113]
### Description

This issue tracks the remaining errors where the TypeScript parser fails to parse a valid TS script. 

* [ ] Abstract / ambient class members #2059 
  * `abstractPropertyBasics`
  * `ambientClassMergesOverloadsWithInterface`
  * `ambientExternalModuleWithoutInternalImportDeclaration`
  * `ambientExternalModuleWithInternalImportDeclaration`
  * `ambiguousCallsWhereReturnTypesAgree`
  * ...
* [x] #2128
  * `aliasUsageInIndexerOfClass`
  * `classImplementingInterfaceIndexer`
  * `classIndexer`
  * `classIndexer5`
  * `declFileClassWithIndexSignature`
  * `indexerA`
* [x] #2122
  * `aliasUsageInTypeArgumentOfExtendsClause`: `extends C<IHasVisualizationModel>` gets incorrectly parsed as `<IHas...>` to 
be a call expression
   * `baseTypeOrderChecking`
   * `baseTypeWrappingInstantiationChain`
   * .. 
* [x]  arrow parsing #2126 
  * `callOfConditionalTypeWithConcreteBranches`
* [x] async arrow with type parameters (@MichaReiser )
* [x] `cast`
  * `commentOnParenthesizedExpressionOpenParen1`
  * `m7Bugs`
  * `arrayBestCommonTypes`
  * `arrowFunctionWithObjectLiteralBody5`
  * `arrowFunctionWithObjectLiteralBody6`
* ~JSX~ (will be covered as a separate umbrella task - https://github.com/rome/tools/issues/2153)
  * `doubleUnderscoreReactNamespace`
  * `jsxFactoryIdentifierAsParameter`
  * `jsxFactoryQualifiedNameWithEs5`
  * `jsxPreserveWithJsInput`
  * `jsxEmitWithAttributes`
  * `jsxFactoryIdentifier`
  * `jsxFactoryQualifiedName`
*  ~decorators~ (non goal for now)
   * `commentOnDecoratedClassDeclaration`
   * `decoratorMetadataForMethodWithNoReturnTypeAnnotation01`
   * `decoratorMetadataNoStrictNull`
   * `decoratorMetadataPromise`
   * `decoratorMetadataConditionalType`
   * `decoratorMetadataOnInferredType`
   * ... 
* [x] #2129
   * `commonJsImportClassExpression`
   * `declarationEmitInferredDefaultExportType2`
   * `exportAssignmentWithoutIdentifier1`
   * `exportEqualsAmd`
   * `exportEqualsUmd`
   * `exportEqualsCommonJs`
   * `exportEqualsProperty`
  * `exportDefaultInterface`
  * `exportDefaultInterfaceAndValue`
  * `moduleAugmentationOfAlias`
* ~labels~
   * `continueTarget3` (tolerable)
   * `escapedIdentifiers`  (blocked by #2035)
* [x] function return type predicate: #2148
   * `controlFlowCommaExpressionAssertionWithinTernary`
   * `requireAssertsFromTypescript`
* [x] TS-JSON options in test files, declaration files, etc. @MichaReiser 
   * `declarationEmitForGlobalishSpecifierSymlink2` 
   * `declarationEmitForGlobalishSpecifierSymlink`
   * `declarationEmitPathMappingMonorepo`
   * `declarationEmitPathMappingMonorepo2`
   * `declarationEmitToDeclarationDirWithDeclarationOption`
   * ...
*  ~RegEx ~(blocked by #2035)
  * `RGI_Emoji_Flag_Sequence`
  * `Emoji_Keycap_Sequence`
  * `castExpressionParentheses`
* [ ] `abstract readonly` vs `readonly abstract` properties
  * `abstractPropertyBasics`
* ~Type Alias parsing~ (blocked by #2035)
  * `parseGenericArrowRatherThanLeftShift.ts`
* [x] `await` and `async` in function expressions/arrow function expressions (maybe related to top level await?)
  * `arrowFunctionWithParameterNameAsync_es2017`
  * `asyncArrowFunction4_es2017`
  * `asyncArrowFunction2_es2017`
  * ...
* [x] private names
  * `privateNameInInExpressionTransform`
* [x] top level await `topLevelAwaitErrors`
* [x] import
  * `cjsImportInES2015`
  * `importAssertion1`
* [ ] [override modifiers](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-3.html#override-and-the---noimplicitoverride-flag)

to be continued once we fixed some of the common errors.

**Comment by MichaReiser** [comment]
I updated the list with the remaining errors. Most parse errors can be assigned one of these four categories

* class Index accessor 
* Decorators (non-goal, but we may want to, at least, skip over them)
* JSX (separate issue)
* `tsconfig.json` inside of test file. We should explore a way to remove the config from the code that we run through the test

**Comment by leops** [comment]
From adding the Prettier test suite in #2151, it seems we're missing the following syntax:
- Cast in assignment positions: `(t as any) = true`
- Typed catch clause: `try {} catch(e: unknown) {}`

Other errors:
- Declaring properties named `get` or `set`: `class Test { declare get: string; }`
- Ambiguous parsing of definitive assignment: `let x<ASI>!function() {};`

**Comment by MichaReiser** [comment]
> * let x<ASI>!function() {};

All these should now have be fixed with the exception of declaring properties. I'll need to revisit the modifier parsing.

**Comment by MichaReiser** [comment]
I'm closing this now and will track the `abstract readonly` and `override` in separate issues.
